diff --git a/package/glibc/glibc.mk b/package/glibc/glibc.mk
index 1eff245588..4e78600c90 100644
--- a/package/glibc/glibc.mk
+++ b/package/glibc/glibc.mk
@@ -91,6 +91,12 @@ GLIBC_CONF_ENV += libc_cv_slibdir=/lib64 libc_cv_rtlddir=/lib
 else
 GLIBC_CONF_ENV += libc_cv_slibdir=/lib32 libc_cv_rtlddir=/lib
 endif
+ifeq ($(BR2_RISCV_ISA_CUSTOM_RVF),)
+RISCV_EXTRA_CONF_OPTIONS += --without-fp
+endif
+ifeq ($(BR2_RISCV_ISA_CUSTOM_RVD),)
+RISCV_EXTRA_CONF_OPTIONS += --without-fp
+endif
 endif
 
 # glibc requires make >= 4.0 since 2.28 release.
@@ -129,7 +135,8 @@ define GLIBC_CONFIGURE_CMDS
 		--disable-werror \
 		--without-gd \
 		--enable-kernel=$(call qstrip,$(BR2_TOOLCHAIN_HEADERS_AT_LEAST)) \
-		--with-headers=$(STAGING_DIR)/usr/include)
+		--with-headers=$(STAGING_DIR)/usr/include \
+		$(RISCV_EXTRA_CONF_OPTIONS))
 	$(GLIBC_ADD_MISSING_STUB_H)
 endef
 
